### PR TITLE
Take custom comparisons into account in set(_Type_)

### PR DESCRIPTION
**Pull request by aarroyoc**
This PR modifies the `set(_Type_)` object to take into account custom comparisons.

Please tell me if you like the solution or if something's missing.

### CODE REVIEW

**Comment by pmoura** [comment]
That was quick 🙂 I will be reviewing the changes tomorrow morning. You name should be added to the new version of the `set/1` object and the date updated. The release notes should also be updated with a short description of the changes and your name as contributor. More tomorrow.

**Comment by pmoura** [comment]
In the test, we should do both `tests(set(integer))` and `tests(set(custom))`.

**Comment by pmoura** [comment]
Also, there's no need of the `catch/3` wrappers or the `iso_compare/3` and `iso_sort/2` predicates as we're defining locally the `compare/3` and `sort/2` predicates and thus overriding those standard Prolog built-in predicates **if** we change `termp` to extend `compringp` **and** update `term` to provide default implementations. Pondering it.

**Comment by pmoura** [comment]
Pushed the changes to the `basic_types` and `types` library to the Logtalk `master` branch. After syncing, you  can now simplify your changes to the `set/1` object by removing the `catch/3` wrappers and the definitions of the `iso_compare/3` and `iso_sort/2` predicates.

**Comment by aarroyoc** [comment]
Thanks for the changes in `basic_types` and `types`. It makes it cleaner now!

**Comment by aarroyoc** [comment]
About tests. I restored the `test(set(integer))`. However, `test(set(custom))` won't be feasible without changes to tests.lgt because a custom object won't be an integer. I think having them separated is fine

File: library/sets/tests_custom.lgt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+:- object(tests_custom,
+	extends(lgtunit)).
+
+    cover(set(reverse_int)).
+
+    % as_set/2 tests
+
+    test(as_set_2_01, deterministic(N == 3)) :-
+        set(reverse_int)::as_set([r1, r2, r3, r4], Set),
+        length(Set, N).
```
Comment (pmoura) [inline library/sets/tests_custom.lgt:71]: We cannot assume that the `length/2` predicate is available as a built-in predicate. We can either write `list::length(…)` or use a `uses/2` directive. Given the other calls, I suggest the former.

Comment (pmoura) [reply library/sets/tests_custom.lgt:71]: P.S. You can set the `portability` flag to `warning` to detect these issues.

Comment (aarroyoc) [reply library/sets/tests_custom.lgt]: TIL about `length/2` not being a built-in

File: library/sets/set1.lgt [hunk 2]
```diff
@@ -19,17 +19,81 @@
...
+
+	compare(>, Term1, Term2) :-
+		_Type_::(Term1 > Term2),!.
+
+	as_set(List, Set) :-
+		sort(List, Set).
+
+	insert([], Element, [Element]).
+	insert([Head| Tail], Element, Set) :-
+		compare(Order, Head, Element),
```
Comment (pmoura) [inline library/sets/set1.lgt:84]: We can improve performance by using an if-then-else in the definition of the `insert/3` predicate without using the custom `compare/3` predicate, which implies creating a choice-point per call followed by its destruction when reaching the cut.

File: library/sets/tests_custom.lgt [hunk 3]
```diff
@@ -0,0 +1,101 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%
+%  This file is part of Logtalk <https://logtalk.org/>
+%  Copyright 1998-2022 Paulo Moura <pmoura@logtalk.org>
```
Comment (pmoura) [inline library/sets/tests_custom.lgt:4]: Add a line here:

```text
%  Copyright 2022 Adrian Arroyo <...@...>
```

Same in the `set1.lgt` file.

File: library/sets/set1.lgt (hunk 2 above)
Comment (aarroyoc) [reply library/sets/set1.lgt]: Good idea!

**Comment by pmoura** [comment]
Thanks for your contribution! There's a portability issue still in the code but is easier to fix after the merge (`Order = <` must be `Order = (<)` for standards compliance and avoiding compilation errors with some Prolog systems).

**Pull request merged**